client/client-multi/client-c/src/platform-impl/kaa_tcp_channel.c

```diff
@@ -0,0 +1,1454 @@
+        kaa_tcp_channel_release_context(kaa_tcp_channel);
+        return KAA_ERR_NOMEM;
+    }
+
+    kaatcp_parser_handlers_t parser_handler;
+
+    parser_handler.connack_handler    = kaa_tcp_channel_connack_message_callback;
+    parser_handler.disconnect_handler = kaa_tcp_channel_disconnect_message_callback;
+    parser_handler.kaasync_handler    = kaa_tcp_channel_kaasync_message_callback;
+    parser_handler.pingresp_handler   = kaa_tcp_channel_pingresp_message_callback;
+    parser_handler.handlers_context   = (void *) kaa_tcp_channel;
+
+    kaatcp_error_t parser_error_code = kaatcp_parser_init(kaa_tcp_channel->parser, &parser_handler);
+    if (parser_error_code) {
+        KAA_LOG_ERROR(logger, parser_error_code, "Failed to initialize Kaa TCP parser");
+        kaa_tcp_channel_release_context(kaa_tcp_channel);
+        return KAA_ERR_TCPCHANNEL_PARSER_INIT_FAILED;
+    }
+
+    /**
+     * Initializes a transport channel interface.
+     */
+    self->context = (void*) kaa_tcp_channel;
+    self->get_protocol_id = kaa_tcp_channel_get_transport_protocol_info;
+    self->get_supported_services = kaa_tcp_channel_get_supported_services;
+    self->release_context = kaa_tcp_channel_release_context;
+    self->sync_handler = kaa_tcp_channel_sync_handler;
+    self->init = kaa_tcp_channel_init;
+    self->set_access_point = kaa_tcp_channel_set_access_point;
+
+    KAA_LOG_TRACE(logger, KAA_ERR_NONE, "Kaa TCP channel created");
+
+    return error_code;
+}
+
+/*
+ * Return transport protocol id constant
+ */
+kaa_error_t kaa_tcp_channel_get_transport_protocol_info(void *context, kaa_transport_protocol_id_t *protocol_info)
+{
+    KAA_RETURN_IF_NIL2(context, protocol_info, KAA_ERR_BADPARAM);
+    kaa_tcp_channel_t *channel = (kaa_tcp_channel_t *) context;
+    *protocol_info = channel->protocol_id;
+    return KAA_ERR_NONE;
+}
+
+/*
+ * Return supported services list
+ */
+kaa_error_t kaa_tcp_channel_get_supported_services(void * context, kaa_service_t **supported_services, size_t *service_count) {
+    KAA_RETURN_IF_NIL3(context, supported_services, service_count, KAA_ERR_BADPARAM);
+    kaa_tcp_channel_t *channel = (kaa_tcp_channel_t *) context;
+
+    *supported_services = channel->supported_services;
+    *service_count = channel->supported_service_count;
+
+    return KAA_ERR_NONE;
+}
+
+/*
+ * Sync specified services list with server side.
+ */
+kaa_error_t kaa_tcp_channel_sync_handler(void *context, const kaa_service_t services[], size_t service_count) {
+    KAA_RETURN_IF_NIL(context, KAA_ERR_BADPARAM);
+    kaa_error_t error_code = KAA_ERR_NONE;
+
+    if (services && service_count > 0) {
+        error_code = kaa_tcp_channel_update_pending_services((kaa_tcp_channel_t *) context, services, service_count);
+        KAA_RETURN_IF_ERR(error_code);
+    }
+
+    return error_code;
+}
+
+/*
+ * Release Kaa TCP channel context
+ */
+kaa_error_t kaa_tcp_channel_release_context(void *context) {
```

Comment:
This is a destroy function. The signature must be:
void kaa_tcp_channel_destroy(void *context)